components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/logout/idpinit/processor/FederatedIdpInitLogoutProcessor.java

```diff
@@ -0,0 +1,632 @@
+     *
+     * @param logoutRequest
+     * @return IdentityResponse.IdentityResponseBuilder.
+     * @throws LogoutClientException Exception occurred due to error in the logout request.
+     * @throws LogoutServerException Exception occurred from IS.
+     */
+    protected IdentityResponse.IdentityResponseBuilder handleOIDCFederatedLogoutRequest(
+            IdentityRequest logoutRequest) throws LogoutClientException, LogoutServerException {
+
+        try {
+            String logoutToken = logoutRequest.getParameter(OIDCAuthenticatorConstants.LOGOUT_TOKEN);
+            if (StringUtils.isBlank(logoutToken)) {
+                throw new LogoutClientException(ErrorMessages.LOGOUT_TOKEN_EMPTY_OR_NULL.getCode(),
+                        ErrorMessages.LOGOUT_TOKEN_EMPTY_OR_NULL.getMessage());
+            }
+
+            if (log.isDebugEnabled()) {
+                log.debug("Handling the OIDC federated IdP Initiated logout request for the obtained logout token: " +
+                        logoutToken);
+            }
+            // Get the claim set from the logout token.
+            SignedJWT signedJWT = SignedJWT.parse(logoutToken);
+            JWTClaimsSet claimsSet = signedJWT.getJWTClaimsSet();
+            // Check for the iss value in claim set.
+            isIssuerClaimExists(claimsSet);
+            // Get the identity provider for the issuer of the logout token.
+            String tenantDomain = logoutRequest.getTenantDomain();
+            IdentityProvider identityProvider = getIdentityProvider(claimsSet.getIssuer(), tenantDomain);
+
+            validateLogoutToken(signedJWT, identityProvider);
+            // Check for sid or sub claims in the logout token.
+            if (isSidClaimExists(claimsSet)) {
+                // Check whether the sid has an entry in the federated authentication session details table.
+                return logoutUsingSid((String) claimsSet.getClaim(OIDCAuthenticatorConstants.Claim.SID));
+            }
+            // Check whether the sub claim is available in the logout token.
+            if (log.isDebugEnabled()) {
+                log.debug("No sid presented in the logout token of the federated idp initiated logout request. Using " +
+                        "sub claim to terminate the sessions for tenant domain: " + tenantDomain);
+            }
+            // Retrieve the federated user id from the IDN_AUTH_USER table.
+            String sub;
+            String userId = null;
+            if (isSubClaimExists(claimsSet)) {
+                sub = claimsSet.getSubject();
+                userId = getUserId(tenantDomain, sub, identityProvider);
+                if (log.isDebugEnabled()) {
+                    log.debug("Trying OIDC federated identity provider initiated logout for the user: " + sub);
+                }
+            }
+            // Check whether the sub claim has a valid user in the IS.
+            return logoutUsingSub(userId);
```

Comment:
Shouldn't this move into the if block.

What happens if sid was present and sub was also not present?
can there be a case like that?